### ISSUE TITLE
Integration with form validation plugins (Example)

### ISSUE THREAD

**Issue by Diaskhan** [bootstrap-vue/bootstrap-vue#171]
Planing a validation mechanism in future for form components ? 
It is a questio  !

**Comment by pi0** [comment]
@Diaskhan Validation won't be directly added to base package as it is not's it's job. But Validation is a SO REQUIRED version and i will add tutorials and examples for integration with vue form validators.  Until then, i suggest you two awesome projects for validation : 
- https://github.com/monterail/vuelidate
- https://github.com/logaretm/vee-validate
Both are active, vue.js 2 compatible and great :)

**Comment by Diaskhan** [comment]
My greetings commander, i love vue js

**Comment by dvatp** [comment]
I disagree with the comment that validation is not part of the "base functionality" of bootstrap-vue.  What I need as a developer is an integrated, fully tested solution of components and validation mechanisms.

I like the vue-strap validation approach, even though it does not currently support all components.  In vue-strap, each component has the ability to accept a "pattern" prop, which is nothing more than a regex indicating the correct content for the field, and exposes a "validation" method which is called behind the scenes by a wrapping "FormValidator" component.

The FormValidator component is configured to bind a "valid" value to some local data.  On large forms groups of components can be wrapped by separate FormValidator components, each binding to their own local data variable indicating the status of that group.  This way I can briefly highlight a section or group of Input components if one of them is invalid to draw the user's eye to the correct section of the page.

My point is that in vue-strap, validating components is as easy as wrapping them with a FormValidator component.  This should also be the case in bootstrap-vue.

**Comment by Diaskhan** [comment]
If will be a form validation, we could named vue strap fully functional framework, thats what U trying to say ?

**Comment by SirLamer** [comment]
Probably also bootstrap-vue should pilot rockets to Mars? Why not?

I will be implementing a validation solution on my own page within the new couple months, and I'll let you know what I come up with. However, that is a thing totally outside of "make page render with Bootstrap 4". This is more of a state management thing. Validation should be performed on the state data, and messages should be displayed to the user when a validation fails, but the act of performing validation is external to how it should be rendered. One person might want to use Bootstrap, another might want to use Foundation.

**Comment by pi0** [comment]
Totally agree with @SirLamer .
@dvatp Validation is not always a simple task like regex checking. The philosophy of web-components is writing *reusable* components to do just what they are meant for. [vuelidate](https://monterail.github.io/vuelidate) is a great library from monterail which is heavily tested for it's specific task which is validation.

**Comment by pi0** [comment]
Also if such integration need necessary API from our components we provide that. If bootstrap + vue + xxx is taking a bad approach it doesn't mean we should take that one too.

**Comment by dvatp** [comment]
You can't deploy bootstrap-vue or any other vue component without validation.  It seems to me the two concepts are married and therefore not outside the scope of the framework.  I know many people will disagree with me, but that's the way I feel.  I don't care about software engineering dogma -- I care about building secure apps quickly and easily, and fully integrating such an essential element as validation into bootstrap-vue would make that possible with a minimum of hassle.

**Comment by SirLamer** [comment]
I will be using vee-validate to perform basic client-side validation, and my client-side solution will also accept a Laravel server-side validation error response via the API and handle that case in a graceful way. This is largely to allow validation like "unique-if" to be performed by the server and integrate in to the client-side interface as fluently as possible. It is also a bit more resilient to future changes. I don't want a nasty script error when server-side validation fails because of a programming change which occurred server-side but was not implemented in the javascript package.

So, basically, each form field will have an object which handles its state, and it will include meta properties like "errors". You just have to make your page render "if there are error message for this form field, display them". And separately keep that error message array up-to-date and/or empty, and don't allow form submission until all error messages are cleared.

**Comment by puzrin** [comment]
@pi0 i've looked validators you mentioned. IMHO it would be nice to have example how to use json-schema validators, instead of custom solutions. Those are standard and don't push users to learn new custom markup every time.

**Comment by Diaskhan** [comment]
As u see gentlmens iam not only one who think that validator or example how to use needed! There are the others !

**Comment by pi0** [comment]
@Diaskhan Thats the only reason this issue is open! If you or any other person have time writing markdown tutorial integration with vue-validatee we will accept it. This is just OSS dude!

**Comment by robcresswell** [comment]
Still open to PRs on this? I could document how I'm using [vuelidate](https://monterail.github.io/vuelidate/), for example.

**Comment by tmorehouse** [comment]
@robcresswell, sure! jut create it as `/docs/markdown/reference/validation/README.md`

**Comment by robcresswell** [comment]
Will do. I'll have some time tomorrow 😃

**Comment by tmorehouse** [comment]
I'll take a look at it when you get it ready, and while I'm on the beach in Mexico 😉

**Comment by robcresswell** [comment]
It's on the list 🙏 Just been a little busy with the run up to Christmas

**Comment by tmorehouse** [comment]
Heheh, no rush... I might be too busy with the tequilas
